src/frame/decompress.rs

```diff
@@ -1 +1,387 @@
+                expected: None,
+            }
+            .into());
+        }
+
+        let max_block_size = frame_info.block_size.get_size();
+        let dst_size = if frame_info.block_mode == BlockMode::Linked {
+            max_block_size * 2 + WINDOW_SIZE
+        } else {
+            max_block_size
+        };
+        #[cfg(feature = "safe-encode")]
+        {
+            self.src.resize(max_block_size, 0);
+            self.dst.resize(dst_size, 0);
+        }
+        #[cfg(not(feature = "safe-encode"))]
+        {
+            self.src.clear();
+            self.dst.clear();
+            self.src.reserve_exact(max_block_size);
+            self.dst.reserve_exact(dst_size);
+            unsafe {
+                self.src.set_len(max_block_size);
+                self.dst.set_len(dst_size);
+            }
+        }
+        self.current_frame_info = Some(frame_info);
+        self.content_hasher = XxHash32::with_seed(0);
+        self.content_len = 0;
+        self.ext_dict_len = 0;
+        self.dst_start = 0;
+        self.dst_end = 0;
+        Ok(required)
+    }
+
+    #[inline]
+    fn read_checksum(r: &mut R) -> Result<u32, io::Error> {
+        let mut checksum_buffer = [0u8; size_of::<u32>()];
+        r.read_exact(&mut checksum_buffer[..])?;
+        let checksum = u32::from_le_bytes(checksum_buffer);
+        Ok(checksum)
+    }
+
+    #[inline]
+    fn check_block_checksum(data: &[u8], expected_checksum: u32) -> Result<(), io::Error> {
+        let mut block_hasher = XxHash32::with_seed(0);
+        block_hasher.write(data);
+        let calc_checksum = block_hasher.finish() as u32;
+        if calc_checksum != expected_checksum {
+            return Err(Error::BlockChecksumError.into());
+        }
+        Ok(())
+    }
+
+    fn read_block(&mut self) -> io::Result<usize> {
+        debug_assert_eq!(self.dst_start, self.dst_end);
+        let frame_info = self.current_frame_info.as_ref().unwrap();
+
+        // Adjust dst buffer offsets to decompress the next block
+        let max_block_size = frame_info.block_size.get_size();
+        if frame_info.block_mode == BlockMode::Linked {
+            if self.dst_start + max_block_size > self.dst.len() {
+                // Output might not fit in the buffer.
+                // The ext_dict will become the last WINDOW_SIZE bytes
+                debug_assert!(self.dst_start >= max_block_size + WINDOW_SIZE);
+                self.ext_dict_offset = self.dst_start - WINDOW_SIZE;
+                self.ext_dict_len = WINDOW_SIZE;
+                // Output goes in the beginning of the buffer again.
+                self.dst_start = 0;
+                self.dst_end = 0;
+            } else if self.dst_start + self.ext_dict_len > WINDOW_SIZE {
+                // Shrink ext_dict in favor of output prefix.
+                let delta = self.ext_dict_len.min(self.dst_start);
+                self.ext_dict_offset += delta;
+                self.ext_dict_len -= delta;
+            }
```

Comment:
what happens in the else case?

Reply:
This is complicated. I added a lot comments, copying the one I added to that block bellow.

>  There's more than WINDOW_SIZE of lookback adding the prefix and ext_dict.
>  Since we have a limited buffer we must shrink ext_dict in favor of the prefix,
>   so that we can fit up to max_block_size bytes between dst_start and ext_dict start.

Reply:
thanks they are really helpful